### PR TITLE
Adopt typed throws in AsyncIteratorProtocol and AsyncSequence

### PR DESCRIPTION
**Pull request by DougGregor**
Introduce a new associated type `Failure` into the two protocols involved
in async sequences, which represents the type thrown when the sequence
fails. Introduce a defaulted `_next(_:)` operations that is isolated to the
given actor (or `nil`) and either throws `Failure` or produces the next
element of the sequence. Provide a default implementation of `next(_:)`
in terms of `next()` that force-casts the thrown error to the `Failure` type,
and a default implementation in terms of `next()` in terms of `next(_:)` that
lets new async iterators only define the new entrypoint.

Drop `@rethrows` from `AsyncSequence` and `AsyncIteratorProtocol`;
this unofficial feature is on its way out.

Introduce special associated type inference logic for the `Failure`
type of an `AsyncIteratorProtocol` conformance when there is no
specific `next(_:)` witness. This inference logic looks at the
witness for `next()`:
* If `next()` throws nothing, `Failure` is inferred to `Never`.
* If `next()` throws, `Failure` is inferred to `any Error`.
* If `next()` rethrows, `Failure` is inferred to `T.Failure`, where
`T` is the first type parameter with a conformance to either
`AsyncSequence` or `AsyncIteratorProtocol`.

The default implementation and the inference rule, together, allow
existing async sequences to continue working as before. Switch the 
contract between the compiler and library for the `async for` loop to
use `next(_:)` rather than `next()` when targeting a new-enough
deployment target.

Give `AsyncSequence` and `AsyncIteratorProtocol` primary associated
types for the element and failure types, which will allow them to be
used more generally with existential and opaque types, i.e.,
`any AsyncSequence<Image, NetworkError>`.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckProtocolInference.cpp [hunk 1]
```diff
@@ -1037,9 +1057,71 @@ Type AssociatedTypeInference::computeFixedTypeWitness(
...
+  if (auto genericSig = dc->getGenericSignatureOfContext()) {
+    for (auto gp : genericSig.getGenericParams()) {
+      // Packs cannot witness associated type requirements.
+      if (gp->isParameterPack())
+        continue;
+
+      if (gp->getName() == assocType->getName())
+        return llvm::None;
+    }
+  }
```
Comment (DougGregor) [inline lib/Sema/TypeCheckProtocolInference.cpp:1079]: This is very sketchy and I want to try eliminating it.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
```
<unknown>:0: error: 'nextElement()' is only available in macOS 9999 or newer
<unknown>:0: error: 'nextElement()' is only available in macOS 9999 or newer
```

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
Ah, SILGen tests failing due to the addition of `@isolated` to SIL. Will fix once the toolchains have built

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS

**Comment by DougGregor** [comment]
The source compatibility issue is real! CareKit is expecting to get a non-throwing async compact flatmap but ends up with a throwing one. This was a known issue, but it isn't clear how widespread it is.

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please build toolchain Windows

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please test windows

File: stdlib/public/Concurrency/AsyncIteratorProtocol.swift [hunk 2]
```diff
@@ -86,13 +86,40 @@ import Swift
 /// checking for cancellation as described above, or in `deinit` if it's
 /// a reference type.
 @available(SwiftStdlib 5.1, *)
-@rethrows
```
Comment (ktoso) [inline stdlib/public/Concurrency/AsyncIteratorProtocol.swift:89]: goodbye 👋 😊

File: stdlib/public/Concurrency/AsyncIteratorProtocol.swift [hunk 3]
```diff
@@ -86,13 +86,40 @@ import Swift
...
+  mutating func next(_ actor: isolated (any Actor)?) async throws(Failure) -> Element?
+}
+
+@available(SwiftStdlib 5.1, *)
+extension AsyncIteratorProtocol {
+  /// Default implementation of `next()` in terms of `next()`, which is
+  /// required to maintain backward compatibility with existing async iterators.
+  @available(SwiftStdlib 5.11, *)
+  @inlinable
+  public mutating func next(_ actor: isolated (any Actor)?) async throws(Failure) -> Element? {
```
Comment (ktoso) [inline stdlib/public/Concurrency/AsyncIteratorProtocol.swift:118]: are we going to do the "conform to just one of them" dance the same as with executors, or different trickery?

**Review by ktoso — APPROVED** [review]
Read through the whole thing so far, not much to add but looking good, heroic effort  Doug :)

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please build toolchain macOS

**Comment by DougGregor** [comment]
@swift-ci please build toolchain Windows

**Comment by DougGregor** [comment]
Lots of spurious failures, so rerunning a number of jobs. Yesterday was a rough day to be a CI system.

File: stdlib/public/Concurrency/AsyncIteratorProtocol.swift (hunk 3 above)
Comment (DougGregor) [reply stdlib/public/Concurrency/AsyncIteratorProtocol.swift:118]: We're relying on the deprecation of the old one to help here, at least from Swift 6 on

Comment (ktoso) [reply stdlib/public/Concurrency/AsyncIteratorProtocol.swift:118]: Okey, so not the the "implement just one of them" -- might be for the best, that was causing a lot of pain in executor rollout, probably not worth it.

**Comment by DougGregor** [comment]
This is under review now: https://forums.swift.org/t/se-0421-generalize-effect-polymorphism-for-asyncsequence-and-asynciteratorprotocol/69662

**Comment by DougGregor** [comment]
I'm going to speculatively merge this to unblock more testing. We can back out the merge or change things depending on how the review goes.

**Pull request merged**